### PR TITLE
chore: add max-fee-check errors and tests to lnd service

### PR DESCRIPTION
**Pull request by vindard**
## Description

This is to explicitly capture some of the logic around max fees when making routeless payments.

In the case of a payment without a fee probe the max fee should always be enforced as this has implications for how we calculate fee reimbursements. If a max fee at the lnd level exceeds what our fee reimbursement expects then this could be a potential attack vector.

Within the lnd service, to correctly catch this we would need to also be aware of sender wallet currencies and price ratios since the actual amount of sats being sent may be less than the actual sender currency amount being deducted (and the corresponding max fee calculation).

**_A new error and test cases were added to ensure we have this included in the lnd service._**

_Note: This shouldn't change any existing maxFee logic. It simply adds additional checks and tests._

### CODE REVIEW

File: src/domain/payments/ln-fees.ts [hunk 1]
```diff
@@ -38,6 +34,36 @@ export const LnFees = (
...
+    senderWalletCurrency: WalletCurrency
+  }) => {
+    const maxFeeAmount = maxProtocolAndBankFee(btcPaymentAmount)
+    const minFeeAmount = priceRatio.convertFromUsd(ONE_CENT)
+    console.log("HERE 0:", {
+      maxFeeToVerify,
+      maxFeeAmount,
+      minFeeAmount,
+      senderWalletCurrency,
+    })
```
Comment (dolcalmi) [inline src/domain/payments/ln-fees.ts:55]: please remove

**Review by dolcalmi — CHANGES_REQUESTED** [review]

File: src/services/lnd/index.ts [hunk 2]
```diff
@@ -624,11 +625,23 @@ export const LndService = (): ILightningService | LightningServiceError => {
...
+    priceRatio: PriceRatio
+    senderWalletCurrency: WalletCurrency
   }): Promise<PayInvoiceResult | LightningServiceError> => {
+    const maxFeeCheck = LnFees().verifyMaxFee({
+      maxFeeToVerify: maxFeeAmount,
+      btcPaymentAmount,
+      priceRatio,
+      senderWalletCurrency,
+    })
+    if (maxFeeCheck instanceof Error) return maxFeeCheck
```
[the remark this answers is not in the document]
Comment (vindard) [reply src/services/lnd/index.ts]: Yea I was thinking this as well when I was writing it. Fixed it here by moving the check to the other side of the lnd service boundary and adding some testing around `maxFee` and payViaPaymentDetails method: https://github.com/GaloyMoney/galoy/pull/2326/commits/31e1b092bc0d2b8814e39674fc0ad85ebc100286 & https://github.com/GaloyMoney/galoy/pull/2326/commits/d4ced76f24b30c5a25b744cd41dc2428189ab5d4

File: test/unit/domain/payments/ln-fees.spec.ts [hunk 3]
```diff
@@ -37,6 +40,113 @@ describe("LnFees", () => {
...
+          priceRatio,
+          senderWalletCurrency: WalletCurrency.Btc,
+        }),
+      ).toBeInstanceOf(MaxFeeTooLargeForRoutelessPaymentError)
+    })
+
+    it("fails for a large Usd maxFee", () => {
+      expect(
+        LnFees().verifyMaxFee({
+          maxFeeToVerify: calc.add(validUsdMaxFeeToVerify, ONE_SAT),
```
[the remark this answers is not in the document]
Comment (vindard) [reply test/unit/domain/payments/ln-fees.spec.ts]: Tests removed with boundary change

File: test/integration/services/lnd/lnd-service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+    it("pays 1 sat with fee at the max limit for BTC wallet", async () => {
+      const { request } = await createInvoice({ lnd: lndOutside1 })
+      const decodedInvoice = decodeInvoice(request)
+      if (decodedInvoice instanceof Error) throw decodedInvoice
+
+      const paid = await lndService.payInvoiceViaPaymentDetails({
+        decodedInvoice,
+        btcPaymentAmount: ONE_SAT,
+        maxFeeAmount: LnFees().maxProtocolAndBankFee(ONE_SAT),
```
[the remark this answers is not in the document]
Comment (vindard) [reply test/integration/services/lnd/lnd-service.spec.ts]: https://github.com/GaloyMoney/galoy/pull/2326#discussion_r1108525504

File: test/unit/domain/payments/ln-fees.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (vindard) [reply test/unit/domain/payments/ln-fees.spec.ts]: Bad naming, does this name change help? https://github.com/GaloyMoney/galoy/pull/2326/commits/beb73802bbc1cd7c5cd355e200a9c008837c6679

File: test/unit/domain/payments/ln-fees.spec.ts [hunk 5]
```diff
@@ -37,6 +40,113 @@ describe("LnFees", () => {
...
+      ).toBeInstanceOf(MaxFeeTooLargeForRoutelessPaymentError)
+    })
+
+    it("fails for a 1 sat large Usd maxFee", () => {
+      expect(
+        LnFees().verifyMaxFee({
+          maxFeeToVerify: calc.add(validUsdMaxFeeToVerify, ONE_SAT),
+          btcPaymentAmount: ONE_SAT,
+          priceRatio,
+          senderWalletCurrency: WalletCurrency.Btc,
```
[the remark this answers is not in the document]
Comment (vindard) [reply test/unit/domain/payments/ln-fees.spec.ts]: Good catch https://github.com/GaloyMoney/galoy/pull/2326/commits/085a4328dc223a604943b0dc010f6bedad603a94

File: test/unit/domain/payments/ln-fees.spec.ts [hunk 6]
```diff
@@ -37,6 +40,113 @@ describe("LnFees", () => {
...
+        }),
+      ).toBeInstanceOf(MaxFeeTooLargeForRoutelessPaymentError)
+    })
+
+    it("fails for a 1 sat large Usd maxFee", () => {
+      expect(
+        LnFees().verifyMaxFee({
+          maxFeeToVerify: calc.add(validUsdMaxFeeToVerify, ONE_SAT),
```
[the remark this answers is not in the document]
Comment (vindard) [reply test/unit/domain/payments/ln-fees.spec.ts]: https://github.com/GaloyMoney/galoy/pull/2326#discussion_r1108530322

File: src/services/lnd/index.ts [hunk 7]
```diff
@@ -651,7 +652,7 @@ export const LndService = (): ILightningService | LightningServiceError => {
       destination: decodedInvoice.destination,
       mtokens: milliSatsAmount.toString(),
       payment: decodedInvoice.paymentSecret as string,
-      max_fee: Number(maxFee),
+      max_fee: maxFee,
```
Comment (dolcalmi) [inline src/services/lnd/index.ts:655]: just a question but not a blocker... is it more safe to include this even if we use routes? or is it completely ignored?

**Review by dolcalmi — APPROVED** [review]

File: src/services/lnd/index.ts (hunk 7 above)
Comment (vindard) [reply src/services/lnd/index.ts:655]: So we have a maxFee passed into probeForRoute which captures this for payViaRoutes... but I was thinking to maybe even remove this in a follow-up PR (following bringing it up at ACE maybe)

**Pull request merged**